File: 10583/main.c
#include <stdio.h>
#include <memory.h>

int main(int argc, char* argv[])
{
   int total;
   int n, m, i, j, tmp;
   int test_case = 1;
   int table[50001];

   while (scanf("%d%d", &n, &m) == 2) {
      if (n == 0 && m == 0) break;
      for (i = 1; i <= n; i++) {
         table[i] = i;
      }
      while (m-- > 0) {
         scanf("%d%d", &i, &j);
         if (table[i] == i) {
            while (table[j] != j) {
               j = table[j];
            }
            table[i] = j;
         } else if (table[j] == j) {
            while (table[i] != i) {
               i = table[i];
            }
            table[j] = i;
         } else {
            while (table[i] != i) {
               i = table[i];
            }
            while (table[j] != j) {
               tmp = table[j];
               table[j] = i;
               j = tmp;
            }
            table[j] = i;
         }
      }
      total = 0;
      for (i = 1; i <= n; i++) {
         if (table[i] == i) {
            ++total;
         }
      }
      printf("Case %d: %d\n", test_case, total);
      ++test_case;
   }

   return 0;
}

File: 10583/Main.java
import java.io.*;
import java.util.*;

public class Main {
   public static void main(String args[]) throws IOException {
      int testCase = 1;
      int[] table = new int[50001];
      String line;
      StringBuilder answer = new StringBuilder();
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((line = in.readLine()) != null) {
         StringTokenizer tokenizer = new StringTokenizer(line);
         int n = Integer.parseInt(tokenizer.nextToken());
         int m = Integer.parseInt(tokenizer.nextToken());
         if (n == 0 && m == 0) break;

         for (int i = 1; i <= n; i++) {
            table[i] = i;
         }

         while (m-- > 0) {
            tokenizer = new StringTokenizer(in.readLine());
            int i = Integer.parseInt(tokenizer.nextToken());
            int j = Integer.parseInt(tokenizer.nextToken());
            if (table[i] == i) {
               while (table[j] != j) {
                  j = table[j];
               }
               table[i] = j;
            } else if (table[j] == j) {
               while (table[i] != i) {
                  i = table[i];
               }
               table[j] = i;
            } else {
               while (table[i] != i) {
                  i = table[i];
               }
               while (table[j] != j) {
                  int tmp = table[j];
                  table[j] = i;
                  j = tmp;
               }
               table[j] = i;
            }
         }
         int total = 0;
         for (int i = 1; i <= n; i++) {
            if (table[i] == i) {
               ++total;
            }
         }
         answer.append("Case ");
         answer.append(Integer.toString(testCase));
         answer.append(": ");
         answer.append(Integer.toString(total));
         answer.append('\n');
         ++testCase;
      }
      System.out.print(answer);
   }
}
